src/main/java/com/google/devtools/build/lib/rules/proto/ProtoCommon.java

```diff
@@ -404,44 +399,57 @@ private static Library createLibraryWithVirtualSourceRootMaybe(
   }
 
   /**
-   * Returns a set of the {@code proto_source_root} collected from the current library and the
-   * specified attribute.
-   *
-   * <p>Assumes {@code currentProtoSourceRoot} is the same as the package name.
+   * Returns a set of pairs of {@code proto_source_root -> [sources]} of `.proto` files that are
+   * exported by a {@code proto_library} target.
    */
-  private static NestedSet<String> getProtoSourceRootsOfAttribute(
-      RuleContext ruleContext, String currentProtoSourceRoot, String attributeName) {
-    NestedSetBuilder<String> protoSourceRoots = NestedSetBuilder.stableOrder();
-    protoSourceRoots.add(currentProtoSourceRoot);
-
-    for (ProtoInfo provider :
-        ruleContext.getPrerequisites(attributeName, Mode.TARGET, ProtoInfo.PROVIDER)) {
-      protoSourceRoots.add(provider.getDirectProtoSourceRoot());
+  private static NestedSet<Pair<PathFragment, ImmutableList<Artifact>>> computeImportableProtos(
+      RuleContext ruleContext, Library library) {
+    NestedSetBuilder<Pair<PathFragment, ImmutableList<Artifact>>> protos =
+        NestedSetBuilder.stableOrder();
+    protos.add(new Pair<>(PathFragment.create(library.getSourceRoot()), library.getSources()));
+    for (ProtoInfo info : ruleContext.getPrerequisites("deps", TARGET, ProtoInfo.PROVIDER)) {
+      protos.addTransitive(info.getExportedProtos());
     }
-
-    return protoSourceRoots.build();
+    return protos.build();
   }
 
   /**
-   * Returns a set of the {@code proto_source_root} collected from the current library and the
-   * direct dependencies.
-   *
-   * <p>Assumes {@code currentProtoSourceRoot} is the same as the package name.
+   * Returns a set of pairs of {@code proto_source_root -> [sources]} of `.proto` files that are
+   * exported by a {@code proto_library} target.
    */
-  private static NestedSet<String> computeStrictImportableProtoSourceRoots(
-      RuleContext ruleContext, String currentProtoSourceRoot) {
-    return getProtoSourceRootsOfAttribute(ruleContext, currentProtoSourceRoot, "deps");
+  private static NestedSet<Pair<PathFragment, ImmutableList<Artifact>>> computeExportedProtos(
+      RuleContext ruleContext, Library library) {
+    NestedSetBuilder<Pair<PathFragment, ImmutableList<Artifact>>> protos =
+        NestedSetBuilder.stableOrder();
+    protos.add(new Pair<>(PathFragment.create(library.getSourceRoot()), library.getSources()));
+    for (ProtoInfo info : ruleContext.getPrerequisites("exports", TARGET, ProtoInfo.PROVIDER)) {
+      protos.addTransitive(info.getExportedProtos());
+    }
+    if (library.getSources().isEmpty()) {
```

Comment:
It appears that `computeExportedProtos` at HEAD does not special-case rules without sources and this one does. Why is that?

Reply:
Actually it does, but it's not so obvious:
https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/proto/ProtoCommon.java;l=67?q=ProtoCommon
(Tested here: https://source.bazel.build/bazel/+/master:src/test/java/com/google/devtools/build/lib/rules/proto/BazelProtoLibraryTest.java;l=53?q=BazelProtoLibraryTest)